### ISSUE TITLE
Lint against non-idiomatic use paths

### ISSUE THREAD

**Issue by EFanZh** [rust-lang/rust-clippy#8038]
### What it does

Warn users about importing item other than modules, types, traits and derive macros.

### Categories (optional)

- Kind: `clippy::style`.

*What is the advantage of the recommended code over the original code*

From <https://doc.rust-lang.org/book/ch07-04-bringing-paths-into-scope-with-the-use-keyword.html#creating-idiomatic-use-paths>:

> Specifying the parent module when calling the function makes it clear that the function isn’t locally defined while still minimizing repetition of the full path.

> when bringing in structs, enums, and other items with use, it’s idiomatic to specify the full path.

The book does not say about how to import macros. Personally, I suggest to warn about importing:

- functions,
- function-like and attribute-like macros,
- consts,
- statics,

while allowing importing:

- derive macros,
- modules,
- traits,
- types.

### Drawbacks

This style might not be suitable for everyone, and should not be enforced by default.

### Example

```rust
mod front_of_house {
    pub mod hosting {
        pub fn add_to_waitlist() {}
    }
}

use self::front_of_house::hosting::add_to_waitlist;

pub fn eat_at_restaurant() {
    add_to_waitlist();
    add_to_waitlist();
    add_to_waitlist();
}
```

Could be written as:

```rust
mod front_of_house {
    pub mod hosting {
        pub fn add_to_waitlist() {}
    }
}

use self::front_of_house::hosting;

pub fn eat_at_restaurant() {
    hosting::add_to_waitlist();
    hosting::add_to_waitlist();
    hosting::add_to_waitlist();
}
```
